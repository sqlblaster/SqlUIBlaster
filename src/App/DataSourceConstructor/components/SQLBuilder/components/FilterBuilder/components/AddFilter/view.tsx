import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { MutateFilterPopover } from '../MutateFilterPopover';
import { AddFilterViewProps } from './props';

export const addFilterPlaceholder = 'Add filters to narrow your answer';

// FIXME: make padding clickable too
export class AddFilter extends React.Component<AddFilterViewProps, any> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddFilterViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const { filters, addFilter, isOpen, handleClose, handleOpen } = this.props;

    return (
      <>
        <AddEntity
          showText={!(filters && filters.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addFilterPlaceholder}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <MutateFilterPopover
            action={addFilter}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
