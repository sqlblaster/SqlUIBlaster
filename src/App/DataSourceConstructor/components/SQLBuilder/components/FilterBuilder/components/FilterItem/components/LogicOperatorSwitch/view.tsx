import { Button, List, ListItem, ListItemText, Typography, withStyles } from '@material-ui/core';
import * as React from 'react';
import { PopoverPaper } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/PopoverPaper';
import { logicOperators } from './logic-operators';
import { LogicOperatorSwitchViewProps } from './props';

export const LogicOperatorSwitch = withStyles({})(
  class extends React.Component<LogicOperatorSwitchViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: LogicOperatorSwitchViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        logicOperator,
        handleLogicOperatorChange,
        isOpen,
        handleOpen,
        handleClose
      } = this.props;

      return (
        <>
          <div style={{ marginLeft: 15 }} ref={this.popoverLauncher} onClick={handleOpen}>
            <Button size='small' style={{ minWidth: 0 }}>
              <Typography color='primary'>{logicOperator}</Typography>
            </Button>
          </div>
          <PopoverPaper
            isOpen={isOpen}
            onClose={handleClose}
            anchorEl={this.popoverLauncher.current}
          >
            <List component='nav' disablePadding={true}>
              {logicOperators.map(lo => (
                <ListItem
                  button={true}
                  key={lo}
                  onClick={handleLogicOperatorChange(lo, handleClose)}
                >
                  <ListItemText inset={true} primary={lo} style={{ padding: 0 }} />
                </ListItem>
              ))}
            </List>
          </PopoverPaper>
        </>
      );
    }
  }
);
