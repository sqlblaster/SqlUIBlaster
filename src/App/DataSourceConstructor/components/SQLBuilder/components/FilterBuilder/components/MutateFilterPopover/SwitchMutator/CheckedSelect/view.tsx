import { Typography, withStyles } from '@material-ui/core';
import CheckIcon from '@material-ui/icons/Check';
import ExpandMore from '@material-ui/icons/ExpandMore';
import * as React from 'react';
import { PopoverPaper } from '../../../../../common/PopoverPaper';
import { CheckedSelectViewProps } from './props';
import { checkedSelectStyles } from './styles';

class View<TItem extends string = string> extends React.Component<
  CheckedSelectViewProps<TItem>
> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: CheckedSelectViewProps<TItem>) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      selectedItem,
      items,
      classes,
      isOpen,
      handleItemSelect,
      showItems,
      handlePopoverClose,
      hasBorder = true
    } = this.props;

    return (
      <>
        <div
          className={`${classes && classes['checked-select']} ${
            hasBorder ? classes && classes['has-border'] : ''
          }`}
          ref={this.popoverLauncher}
          onClick={showItems}
        >
          <Typography className={classes && classes['selected-item-value']}>
            {selectedItem}
          </Typography>
          {hasBorder && <ExpandMore fontSize='small' />}
        </div>
        <PopoverPaper
          isOpen={isOpen}
          onClose={handlePopoverClose}
          anchorEl={this.popoverLauncher.current}
        >
          <div className={classes && classes['select-items']}>
            {items &&
              items.map(item => {
                const selected = item === selectedItem;

                return (
                  <div
                    className={`${classes && classes.item}${
                      selected ? ` ${classes && classes['selected-item']}` : ''
                    }`}
                    key={item}
                    onClick={handleItemSelect(item)}
                  >
                    <CheckIcon
                      fontSize='small'
                      className={classes && classes['check-icon']}
                      style={{
                        ...(!selected ? { color: 'transparent' } : {})
                      }}
                    />
                    <Typography className={classes && classes['item-text']}>
                      {item}
                    </Typography>
                  </div>
                );
              })}
          </div>
        </PopoverPaper>
      </>
    );
  }
}

/**
 * This type casting was necessary to do,
 * cuz wrapping generic component to HOC
 * eliminates generic part as an outcome
 */
export const CheckedSelect: typeof View = withStyles(checkedSelectStyles)(
  View
) as any;
