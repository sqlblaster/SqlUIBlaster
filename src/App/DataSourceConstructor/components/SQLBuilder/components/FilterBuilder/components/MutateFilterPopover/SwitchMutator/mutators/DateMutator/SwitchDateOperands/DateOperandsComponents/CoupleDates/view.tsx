import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { DatePickerByOperandsType } from '../../DatePickerByOperandsType';
import { CoupleDateViewProps } from './props';
import { coupleDateStyles } from './styles';

export const CoupleDate = withStyles(coupleDateStyles)((({
  operatorView,
  operands,
  pickedVariables,
  classes,
  handleVariablePick,
  handleOperandsChange
}) => {
  return (
    <>
      <div className={classes.root}>
        {operatorView}
        <div className={classes.dates}>
          <div className={classes.date} data-testid='between-operand-1'>
            <DatePickerByOperandsType
              operands={operands[0]}
              onOperandsChange={handleOperandsChange(0)}
              variable={pickedVariables[0]}
              onVariablePicked={handleVariablePick(0)}
              limits={
                operands[1].dateType === 'Fixed date'
                  ? {
                      maxDate: operands[1].date,
                      maxDateMessage:
                        'Couldn\'t set time, because selected date is later than second date'
                    }
                  : undefined
              }
            />
          </div>
          <div className={classes.date} data-testid='between-operand-2'>
            <DatePickerByOperandsType
              operands={operands[1]}
              onOperandsChange={handleOperandsChange(1)}
              variable={pickedVariables[1]}
              onVariablePicked={handleVariablePick(1)}
              limits={
                operands[0].dateType === 'Fixed date'
                  ? {
                      minDate: operands[0].date,
                      minDateMessage:
                        'Couldn\'t set time, because selected date is earlier than first date'
                    }
                  : undefined
              }
            />
          </div>
        </div>
      </div>
    </>
  );
}) as React.FC<CoupleDateViewProps>);
