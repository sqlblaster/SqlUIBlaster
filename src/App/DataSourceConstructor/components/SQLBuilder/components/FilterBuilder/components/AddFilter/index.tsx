import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  AddFilterDispatchProps,
  AddFilterProps,
  AddFilterStateProps,
  FBMapDispatchToProps
} from './props';
import { AddFilter as View } from './view';

export const AddFilter = connect<
  AddFilterStateProps,
  AddFilterDispatchProps,
  {},
  State
>(
  ({ filters }) => ({
    filters
  }),
  FBMapDispatchToProps
)(
  class extends React.Component<AddFilterProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public render() {
      const { filters, addFilter } = this.props;

      return (
        <AddFilter.ViewWithPopoverManagement
          filters={filters}
          addFilter={addFilter}
        />
      );
    }
  }
);
