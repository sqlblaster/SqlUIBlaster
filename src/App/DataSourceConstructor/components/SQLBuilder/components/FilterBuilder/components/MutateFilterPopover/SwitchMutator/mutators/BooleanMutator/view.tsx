import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../CheckedSelect';
import { BooleanOperator, booleanOperators } from './operators';
import { BooleanMutatorViewProps } from './props';
import { booleanMutatorStyles } from './styles';

export const BooleanMutator = withStyles(booleanMutatorStyles)(
  class extends React.Component<BooleanMutatorViewProps> {
    public render() {
      const {
        operator,
        handleOperatorSelect,
        titleComponent,
        classes,
        submitMutationComponent
      } = this.props;

      return (
        <div className={classes.root}>
          <div
            className={ `${classes['title-and-operator']} ${
              classes['common-padding']
            }` }
          >
            <div className={classes.title}>{titleComponent}</div>
            <div className={classes.operator}>
              <CheckedSelect<BooleanOperator>
                selectedItem={operator}
                items={booleanOperators}
                onSelect={handleOperatorSelect}
              />
            </div>
          </div>
          <hr className={classes.hr} />
          <div
            className={ `${classes['mutator-action']} ${
              classes['common-padding']
            }` }
          >
            {submitMutationComponent}
          </div>
        </div>
      );
    }
  }
);
