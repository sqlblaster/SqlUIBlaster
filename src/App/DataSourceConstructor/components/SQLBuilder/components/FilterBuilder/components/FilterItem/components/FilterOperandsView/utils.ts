import { Operation } from '../../../MutateFilterPopover/Operation';
import { DateOperation } from '../../../MutateFilterPopover/SwitchMutator/mutators/DateMutator/DateOperation';
import { CoupleOperandsViewProps } from './components/CoupleOperandsView';
import { SingleOperandViewProps } from './components/SingleOperandView';
import {
  formatCoupleDateOperands,
  formatNearDateOperands,
  formatSingleDateOperands
} from './date-formatter';

export const hasOperands = (operator: Operation['operator']): boolean => {
  switch (operator) {
    case 'Is empty':
    case 'Is not empty':
    case 'Is true':
    case 'Is false':
      return false;
    default:
      return true;
  }
};

export const getFormattedDateOperand = (operation: DateOperation): string => {
  switch (operation.operator) {
    case 'Previous':
    case 'Next':
      return formatNearDateOperands(operation.operands, operation.operator);
    case 'Before':
    case 'After':
    case 'On':
      return formatSingleDateOperands(operation.operands);
    default:
      return '';
  }
};

export const getFormattedCoupleOperands = (
  operation: Operation
): CoupleOperandsViewProps['operands'] => {
  if (operation.type === 'Date' && operation.operands instanceof Array) {
    return formatCoupleDateOperands(operation.operands);
  } else if (
    operation.type === 'Number' &&
    operation.operands instanceof Array
  ) {
    return operation.operands;
  } else {
    throw new Error('expected between operands');
  }
};

export const getFormattedSingleOperand = (
  operation: Operation
): SingleOperandViewProps['operand'] => {
  if (
    hasOperands(operation.operator) &&
    !(operation.operands instanceof Array) &&
    hasOperands(operation.operator)
  ) {
    switch (operation.type) {
      case 'Foreign': {
        const record = operation.operands;

        return record && record.title ? record.title : '';
      }
      case 'Date':
        return getFormattedDateOperand(operation);
      case 'Number':
      case 'String':
      default:
        return operation.operands as number | string;
    }
  } else {
    throw new Error('expected singular operand');
  }
};
