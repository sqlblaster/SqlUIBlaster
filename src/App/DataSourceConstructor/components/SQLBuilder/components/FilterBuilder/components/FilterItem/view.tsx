import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnItem } from '../../../common/SequentialColumnView';
import { MutateFilterPopover } from '../MutateFilterPopover';
import { Brackets } from './components/Brackets';
import { FilterOperandsView } from './components/FilterOperandsView';
import { LogicOperatorSwitch } from './components/LogicOperatorSwitch';
import { operatorSuffixMap } from './operator-suffix-map';
import { FilterItemViewProps } from './props';
import { filterItemStyles } from './styles';

export const FilterItem = withStyles(filterItemStyles)(
  class extends React.Component<FilterItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: FilterItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        filter,
        index,
        updateFilter,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleFilterRemoval,
        handleLogicOperatorSwitch,
        handleBracketsCountChange
      } = this.props;

      const {
        id,
        column,
        operation,
        variables,
        logicOperator,
        openingBracketsCount,
        closingBracketsCount
      } = filter;

      return (
        <>
          <div className={classes.root + (isOpen ? ` ${classes.open}` : '')}>
            {index !== 0 && (
              <LogicOperatorSwitch
                logicOperator={logicOperator}
                onLogicOperatorSwitch={handleLogicOperatorSwitch}
              />
            )}
            <div className={classes.expression}>
              <Brackets
                bracketsCount={openingBracketsCount}
                onChange={handleBracketsCountChange}
                openBracket={true}
              />
              <div
                className={classes.operation}
                onClick={handleOpen}
                ref={this.popoverLauncher}
              >
                <SequentialColumnItem
                  {...column}
                  suffix={operatorSuffixMap[operation.operator]}
                />
                <FilterOperandsView
                  operation={operation}
                  variables={variables}
                />
              </div>
              <RemoveButton onClick={handleFilterRemoval(id)} />
              <Brackets
                bracketsCount={closingBracketsCount}
                onChange={handleBracketsCountChange}
                openBracket={false}
              />
            </div>
          </div>
          {isOpen && (
            <MutateFilterPopover
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateFilter}
              onClose={handleClose}
              filter={filter}
            />
          )}
        </>
      );
    }
  }
);
