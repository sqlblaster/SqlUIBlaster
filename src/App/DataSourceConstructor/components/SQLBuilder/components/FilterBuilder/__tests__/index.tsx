import { fireEvent, getByText as getByTextOn } from 'react-testing-library';
import { backIconPlaceholder } from 'src/__tests__/utils/mock-placeholders';
import { regexWord } from 'src/components/query-builder/App/DataSourceConstructor/test-utils/utils';
import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { selectAColumn } from '../../common/ColumnSelector/__tests__/utils/select-a-column';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import {
  clickAddAFilter,
  clickAddMoreFilter
} from '../components/AddFilter/__tests__/utils/click-add-a-filter';
import { addFilterPlaceholder } from '../components/AddFilter/view';
import { clickAddTheFilter, clickUpdateTheFilter } from './utils/filter-utils';

describe('Filters Builder', () => {
  it(`should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter`, () => {

    const { getByText, queryByText, queryAllByText, getAllByText } = render();

    const columns = selectATable();

    // shows the add filter placeholder
    expect(getByText(addFilterPlaceholder)).toBeInTheDocument();

    // add filter
    clickAddAFilter();
    selectAColumn(columns[3]);
    clickAddTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot();

    // hides logic operator before first operation
    expect(queryByText('and')).not.toBeInTheDocument();

    // hides the add filter placeholder
    expect(queryByText(addFilterPlaceholder)).not.toBeInTheDocument();

    // update filter to the different filter type
    fireEvent.click(getByText(/is equal to/i));
    fireEvent.click(getByText(backIconPlaceholder));
    selectAColumn(columns[2]);
    clickUpdateTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot();

    // add one more filter
    clickAddMoreFilter();
    selectAColumn(columns[1]);
    clickAddTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot();

    // still hides logic operator before first operation
    expect(queryAllByText(/and/i).length).toBe(1);

    // change logic operator's value
    fireEvent.click(getByText(regexWord('and')));
    fireEvent.click(getByText(regexWord('or')));

    expect(getGeneratedQuery()).toMatchSnapshot();

    // remove filter
    fireEvent.click(getAllByText(/clear/i)[0]);

    expect(getGeneratedQuery()).toMatchSnapshot();
  });
});
