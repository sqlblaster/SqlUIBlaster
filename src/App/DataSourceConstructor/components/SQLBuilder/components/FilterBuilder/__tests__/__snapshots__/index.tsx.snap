// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Builder should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" = 0
LIMIT 2000
"
`;

exports[`Filters Builder should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter 2`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) >= date_trunc('Days', (now() + interval '-30 Days')::timestamp)
  AND date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) <= date_trunc('Days', (now() + interval '-1 Days')::timestamp)
LIMIT 2000
"
`;

exports[`Filters Builder should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter 3`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) >= date_trunc('Days', (now() + interval '-30 Days')::timestamp)
  AND date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) <= date_trunc('Days', (now() + interval '-1 Days')::timestamp)
  AND \\"orders1\\".\\"Details\\"::text = 'Detail1'
LIMIT 2000
"
`;

exports[`Filters Builder should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter 4`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) >= date_trunc('Days', (now() + interval '-30 Days')::timestamp)
  AND date_trunc('Days', (\\"orders1\\".\\"Created At\\")::timestamp) <= date_trunc('Days', (now() + interval '-1 Days')::timestamp)
  OR \\"orders1\\".\\"Details\\"::text = 'Detail1'
LIMIT 2000
"
`;

exports[`Filters Builder should be able to
        shows the add filter placeholder,
        add a filter,
        hide the add filter placeholder,
        update filter to the different filter type,
        add one more filter,
        hide logic operator before first operation,
        change logic operator's value,
        remove that filter 5`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Details\\"::text = 'Detail1'
LIMIT 2000
"
`;
